### PR TITLE
Fully declarative modules

### PR DESCRIPTION
**Pull request by shlevy**
Building on top of #2540, this allows modules to specify the path of the nixpkgs lib to be used for the evaluation. In addition, if a nix with https://github.com/NixOS/nix/commit/c273c15cb13bb86420dda1e5341a4e19517532b5 is used, they can also specify the scope to use when importing modules, which in conjunction with https://github.com/NixOS/nix/commit/62a6eeb1f3da0a5954ad2da54c454eb7fc1c6e5d allows the nix path to be set for the module evaluation.

### CODE REVIEW

**Comment by aszlig** [comment]
@edolstra, @rbvermaa: Any chance of finally getting this in? I frequently stumble on the status quo doing workarounds like [this one](https://gist.github.com/aszlig/f355ff7cf5b9779b7be0#file-spectrum2-nix-L132-L139). :unamused:

File: nixos/modules/system/boot/systemd.nix [hunk 1]
```diff
@@ -1,7 +1,7 @@
 { config, lib, pkgs, utils, ... }:
 
-with lib;
 with utils;
+with lib;
```
Comment (wmertens) [inline nixos/modules/system/boot/systemd.nix]: If the order is that important, perhaps a comment would be warranted. Is this picking `lib` from `utils`?

**Comment by wmertens** [comment]
I think I like this but I don't fully understand it. It would be great if there were some comments in the code explaining the module evaluation flow and what `__internal` should be used for.

The changed modules do seem a little cleaner.

How about we just go for 5 :+1:'s since there seems to be no vetoing? At 5 we merge.

:+1: from me while imploring for comments.

**Comment by shlevy** [comment]
Someone else will have to pick this up if it is wanted, I'm done with it

**Comment by geerds** [comment]
It's sad to see that nobody knows whether a merge is acceptable or not. We need to organize our project in a better way :pensive: 

@edolstra Is there a specific reason why you didn't comment this PR yet? Is it a time related issue? If so, it would be great if we had one more person who's responsible for technical decisions

**Comment by copumpkin** [comment]
What's sadder to me than this particular instance is to see talented contributors basically giving up on the entire process as a result of things like this. Agreed that we need to come up with a more formal evaluation structure for things like this, because I don't think the project can afford to lose expertise like @shlevy's.

**Comment by copumpkin** [comment]
One possibility, given limited time/energy, is for @edolstra to tell certain people that he blanket-trusts their judgment on various sorts of things. Thus if @shlevy wanted to make a change like this to the module system, he might be asking for advice/comments, but not for permission. I realize that this PR (and the others that have died on the vine like it) was phrased a request for comments, but I think there's a fairly widespread sense (perhaps unjustified!) that we shouldn't be making deep changes to things like this without @edolstra's approval.

I don't personally know who's credible in what areas of nix, but I'm sure it's easy to figure out, and delegating some of that authority seems like a necessary requisite for this project to scale further, as Eelco has limited time and can't chime in on everything.

**Comment by peti** [comment]
The way I read the comments from @shlevy, @aszlig, @copumpkin, @edwtjo, and @wmertens in this thread feels like you all agree that this PR is useful and that it should be merged. At the same time, it's curious that *everyone of you* has the ability to press the merge button. So why don't you? If you want that patch in Nixpkgs, then merge it! Don't wait for @edolstra to give you permission. His lack of commentary here is a clear indication that he doesn't have an opinion on this matter, so it's pointless to wait for him to make a decision. Clearly he won't. Someone else has to do it.

**Comment by wmertens** [comment]
I'd have pressed that merge button if I fully understood the code, however
I don't, yet :)
What I'd like to understand is if it means that modules will behave
differently for people that define modules outside of the tree.

**Comment by copumpkin** [comment]
@peti, that boils down to my point earlier:

>  but I think there's a fairly widespread sense (perhaps unjustified!) that we shouldn't be making deep changes to things like this without @edolstra's approval.

I think I'm credible in some things and not in others. I like what this PR claims to do, but I don't know if it does it in a good way, so I don't feel like it would be right for me to merge it. I do trust @shlevy's judgment in general, but I also don't know if I'm credible in knowing who to trust, so the same concern applies :smile: 

Anyway, I'd be happy with a decision structure like the one you're proposing, but I think there should be an explicit statement somewhere (from @edolstra or @rbvermaa) that that's how they want to run the project. I haven't seen such a thing (is there one?), so I default to being fairly cautious, and have observed quite a few other people doing the same.

**Comment by peti** [comment]
I cannot speak for @edolstra or @rbvermaa, obviously, but my personal interpretation of the project's policy is that every "collaborator" is free to commit any change (after some reasonable grace period for others to review it). There's no "ownership" of certain parts of Nixpkgs. If you got commit rights then go ahead and commit!

However, if you commit a change that causes trouble for someone else, then the general expectation is that the committer takes responsibility for remedying those issues. This doesn't mean that you have to personally fix every single problem in arcane parts of the package tree, but there should be a sense that the committer takes those issues seriously and makes an effort to get everything back into a working state. If in doubt, that means reverting the entire change and going back to the drawing board.

Personally, I won't press the merge button because I don't understand what exactly this patch does and I don't want to be responsible for breaking other people's system in a way that I can neither predict nor fix. My impression is that many others feel the same way. So everyone seems to be looking at @edolstra, because we expect him to understand everything, but that's not a reasonable expectation. My guess is that Eelco has no clue about the potential side-effects of this change either (otherwise he would have pointed them out).

**Comment by copumpkin** [comment]
@peti that makes sense to me. Perhaps we should just write down some brief contributor guidelines? Then someone (perhaps even @shlevy, invigorated by newfound enthusiasm stemming from the new guidelines and the ensuing carte blanche) can go revive all these dead PRs. Flowers, rainbows, unicorns, and narwhals will follow :tulip: :rainbow: :elephant :whale:

**Comment by shlevy** [comment]
@peti before I opened these PRs I discussed the ideas with @rbvermaa and @edolstra and they said they would need to approve before merging.

**Comment by shlevy** [comment]
Anyway, at this point we are not using these changes and have plans to move from the module system in general, so this particular work is moot for us.

**Comment by copumpkin** [comment]
@shlevy "us" there being Zalora, right? The changes might still be relevant to the rest of the Nix community, if someone (presumably not you, then) is willing to put in the work.

**Comment by shlevy** [comment]
Yes, that's what I neant

**Comment by nbp** [comment]
@shlevy , I will try to review and comment / approve this pull request by the end of next week.

**Comment by aszlig** [comment]
nbp: Thanks a bunch, I'll review/comment it as well as soon as I'm back home =)

**Comment by nbp** [comment]
I will continue to review these changes later this week.

**Comment by edolstra** [comment]
@nbp I don't really see the use case for this change, which adds a lot of complexity (such as the whole `scopedImport` stuff). It's apparently to override Nixpkgs `lib` for module evaluation, but I can't really see where that's useful...

**Comment by shlevy** [comment]
This does more than just override `lib` for module evaluation, but that specific change was added so that we could specify a particular nixpkgs revision in the deployment expressions and still use nixops regardless of what nixpkgs was in NIX_PATH (as long as it was new enough to have these changes of course)

**Comment by nbp** [comment]
The pull-request #6794 is taking over the work made on #2540 , which is included by these changes. This pull-request was only for the the work done in commit 084a017, which sounds like adding some complexity for a small benefit which is yet to be seen.

I do not think the module system is changing that frequently to require something which include a specific version of it.  I understand the need for using a different of `nixpkgs`, but this is orthogonal to the version used for the evaluation of the module system, especially since the `pkgs` argument is computed based on the expressions which are in the module / submodule.

**Comment by shlevy** [comment]
FWIW, it's not orthogonal, because `eval-modules.nix` hard-codes the default module list. So you can override the packages used but not the list of modules.

**Comment by nbp** [comment]
Indeed, this is something that I want to address at some point, in order to add logic at the level of the imports such as one module can claim to replace another one, but the `scopedImport` is not easy to understand.

**Pull request closed without merging**